File: 160-intersection-of-two-linked-lists/intersection-of-two-linked-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
    
    def getIntersectionNode(
        self, headA: ListNode, headB: ListNode
    ) -> Optional[ListNode]:

        cur1, cur2, n1, n2 = headA, headB, 1, 1
        while cur1.next:
            n1 += 1
            cur1 = cur1.next
        while cur2.next:
            n2 += 1
            cur2 = cur2.next

        if cur1 != cur2:
            return None

        cur1, cur2 = headA, headB
        if n1 > n2:
            while n1 != n2:
                n1 -= 1
                cur1 = cur1.next
        elif n1 < n2:
            while n1 != n2:
                n2 -= 1
                cur2 = cur2.next

        while cur1 != cur2:
            cur1 = cur1.next
            cur2 = cur2.next

        return cur1

File: 160-intersection-of-two-linked-lists/intersection-of-two-linked-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode(int x) {
 * val = x;
 * next = null;
 * }
 * }
 */

public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        ListNode temp1 = headA, temp2 = headB;
        int n1 = 1, n2 = 1;

        while (temp1.next != null) {
            n1 += 1;
            temp1 = temp1.next;
        }
        while (temp2.next != null) {
            n2 += 1;
            temp2 = temp2.next;
        }

        if (temp1 != temp2)
            return null;
        temp1 = headA;
        temp2 = headB;

        if (n1 > n2) {
            while (n1 != n2) {
                n1 -= 1;
                temp1 = temp1.next;
            }
        } else if (n1 < n2) {
            while (n1 != n2) {
                n2 -= 1;
                temp2 = temp2.next;
            }
        }

        while (temp1 != temp2) {
            temp1 = temp1.next;
            temp2 = temp2.next;
        }
        return temp1;
    }
}
